### DISCUSSION TITLE
How can I pass a Param via CLI

### DISCUSSION THREAD

**Discussion by aamster** [apache/airflow#28896] [Q&A]
I have a dag like

```
import datetime

from airflow.models import Param
from airflow.models.dag import dag


@dag(
    dag_id="foo",
    schedule=None,
    params={'month': Param()},
    start_date=datetime.datetime.now()
)
def foo():
    pass


foo()

```

I am trying to pass this param via CLI. I tried passing it like 
```
airflow dags test foo --conf '{"month": "2022-01"}'
```

gives
```
airflow.exceptions.ParamValidationError: Invalid input for param month: No value passed and Param has no default value
```

This is not answered on doc page https://airflow.apache.org/docs/apache-airflow/stable/concepts/params.html

**Comment by potiuk** [comment]
My guess is that  you might be using old airflow version where test command did not support conf. You forgot to mention your version but it's been added in 2.4.0 https://github.com/apache/airflow/pull/25900

**Reply by aamster** [reply]
I am using the latest 2.5.0

**Reply by potiuk** [reply]
No idea then. Maybe someone else can help. Or you can just debug it.

**Reply by aamster** [reply]
Unfortunately I tried for hours to debug and it couldn't figure it out

**Reply by potiuk** [reply]
I thnk if you pass here a fully reproducible example you try and full stack trace of what you get, it might help others to help you to solve an issue you have. Just make sure it is as easy as copying your example to reproduce.

**Reply by potiuk** [reply]
Just help others to help you and understand that people here are mostly volunteers who help others in their free time.

**Reply by aamster** [reply]
Sure, here is the complete code
```
import datetime
import os
import random
from collections import defaultdict
from pathlib import Path

import pandas as pd
from airflow.decorators import dag, task
from airflow.models import Param
from airflow.operators.python import get_current_context


@dag(
    dag_id="artist_top_list",
    schedule=None,
    catchup=False,
    params={'month': Param()},
    start_date=datetime.datetime(year=2022, month=1, day=1)
)
def artist_top_list():
    @task
    def parse_input():
        context = get_current_context()
        month = context['params']['month']
        month = datetime.datetime.strptime(month, '%Y-%m')
        date_interval = pd.date_range(
            start=month,
            end=datetime.datetime(year=month.year, month=month.month + 1,
                                  day=1))
        return [str(x) for x in date_interval.to_list()]

    @task
    def streams(date: str):
        date = datetime.datetime.strptime(date, '%Y-%m-%d %H:%M:%S')
        path = date.strftime('data/streams_%Y_%m_%d_faked.tsv')
        os.makedirs(Path(path).parent, exist_ok=True)
        with open(path, 'w') as output:
            for _ in range(1000):
                output.write('{} {} {}\n'.format(
                    random.randint(0, 999),
                    random.randint(0, 999),
                    random.randint(0, 999)))
        return path

    @task
    def aggregate_artists(streams_paths):
        artist_count = defaultdict(int)

        for s in streams_paths:
            with open(s) as f:
                for line in f:
                    _, artist, track = line.strip().split()
                    artist_count[artist] += 1

        context = get_current_context()
        month = context['params']['month']

        out = f"data/artist_streams_{month}.tsv"
        with open(out, 'w') as f:
            for artist, count in artist_count.items():
                f.write('{}\t{}\n'.format(artist, count))
        return out

    @task
    def top_10_artists(artist_agg_path):
        df = pd.read_csv(artist_agg_path, sep='\t')
        df.columns = ['artist_id', 'count']
        top10 = df.sort_values(by='count', ascending=False)[:10].copy()

        context = get_current_context()
        start_date = context['params']['month']
        start_date = datetime.datetime.strptime(start_date, '%Y-%m')
        end_date = datetime.datetime(
            year=start_date.year, month=start_date.month + 1, day=1)
        end_date = end_date.strftime('%Y-%m')
        top10['start'] = start_date
        top10['end'] = end_date

        out = f"data/top_artists_{start_date}.tsv"
        top10.to_csv(out, index=False)
        return out

    date_interval = parse_input()
    streams_paths = streams.expand(date=date_interval)
    artist_agg_path = aggregate_artists(streams_paths)
    top10 = top_10_artists(artist_agg_path=artist_agg_path)


artist_top_list()


if __name__ == '__main__':
    artist_top_list().test(run_conf={'month': '2022-01'})

```

traceback

```
Traceback (most recent call last):
  File "/Users/adam.amster/venvs/airflow/lib/python3.10/site-packages/airflow/models/dagbag.py", line 437, in _process_modules
    dag.validate()
  File "/Users/adam.amster/venvs/airflow/lib/python3.10/site-packages/airflow/models/dag.py", line 667, in validate
    self.params.validate()
  File "/Users/adam.amster/venvs/airflow/lib/python3.10/site-packages/airflow/models/param.py", line 230, in validate
    raise ParamValidationError(f"Invalid input for param {k}: {ve}") from None
airflow.exceptions.ParamValidationError: Invalid input for param start_date: No value passed and Param has no default value

```

**Reply by potiuk** [reply]
As I thought. You have an obvious mistake in your DAG. you have no "month" parameter. You have `start_date` and you are passing `month` instead.

**Reply by aamster** [reply]
Apologies. I typed `start_date` instead of `month` into the comment above. I updated the code in the comment. I get the same error with `month`

**Reply by potiuk** [reply]
If you still has a problem of this kind - I advice you to distill it to really simple reproducible one-liner and post it as reproduction here to help. That will often lead you through the process when while distilling it down you will find out the problem yourself and avoid posting it at all (and you will learn help others to spend as little time on your typos /mistakes as possible).

**Reply by potiuk** [reply]
I thin if you follow the process I described above ^^  to distill things down it will be much more useful. That's what I usually do when debuggging problems - this is an approach that is mindful to time of people that are trying to help you.

**Reply by aamster** [reply]
I tried to post a minimal example in the description

**Reply by aamster** [reply]
Anyways I've deleted `params={'month': Param()}` from the dag. It seems to serve no purpose as far as I can see. Once removed I can access using
```
        context = get_current_context()
        month = context['params']['month']
```
by passing `--conf '{"month": "2022-01"}'` to the CLI. So it seems `params` serves no purpose. 

(although I'd still like to use it since it enforces arguments passed into the workflow)

**Reply by potiuk** [reply]
> I tried to post a minimal example in the description

When I copy & paste it it does not work so it's not reproducible.

> by passing --conf '{"month": "2022-01"}' to the CLI. So it seems params serves no purpose.

You seem to jump to conclusions very quickly.

> (although I'd still like to use it since it enforces arguments passed into the workflow)

Yes if you post a minimum-reproducible example with single task that can be copy&pasted - without all the overhead and potential typos in your full dag. Someone might be able to point out where you've made a mistake, because this is the purpose of params.

**Reply by aamster** [reply]
@potiuk I have edited it to make it fully reproducible.

**Reply by aamster** [reply]
Found out the issue and created a bug report https://github.com/apache/airflow/issues/28940

**Reply by potiuk** [reply]
Cool. I just returned from a weekend and was about to take a look, but you beat me to it. Marked it as `good first issue` and hopefully someone will fix it (but I also encourage you to take a look and maybe you will be able ot fix it and become a contributor).
